### PR TITLE
Start type-checking redux-persist code

### PR DESCRIPTION
**Pull request by chrisbobbe**
(draft)

### CODE REVIEW

**Comment by gnprice** [comment]
OK, I've just pushed a version of this that completes type-checking the rest of the library. @chrisbobbe please take a look!

The type-checking is complete at the commit "Cut .js.flow". In the commits after that, we go on to make some refactors that were best done with the help of types -- in particular, converting `getStoredState` to present a Promise-based API rather than callback-based.

(Some of the early commits in this branch were written by me and Chris pairing, several weeks ago. I don't recall exactly which commits those were.)

**Review by chrisbobbe — COMMENTED** [review]
Thanks! LGTM except one small question below.

File: src/third/redux-persist/persistStore.js [hunk 1]
```diff
@@ -26,96 +26,96 @@ export default function persistStore<
...
-        // Caution: this same expression would give a different
-        // value if it were run after the `store.dispatch` line,
-        // because redux-persist-migrate's store enhancer
-        // `createMigration` mutates `migrations.version` in the
-        // action's payload (see `realVersionSetter` in
-        // redux-persist-migrate).
-        //
-        // $FlowFixMe[prop-missing] This can indeed be missing; that's why we check
-        const prevVersion: mixed = restoredState.migrations?.version;
+          resolve(state);
```
[the remark this answers is not in the document]
Comment (gnprice) [reply src/third/redux-persist/persistStore.js]: Ah, that's an oversight in this commit -- I should put a `return` after the `reject`.

> It seems unwise for Promise implementations to let the resolve call change the Promise's state after the reject call. I definitely hope that the implementations we're using don't do that.

Yeah, the spec does end up defining this. From your link, the callback to the Promise constructor gets invoked like so:
> 8. Let resolvingFunctions be CreateResolvingFunctions(promise).
> 9. Let completion be Call(executor, undefined, « resolvingFunctions.[[Resolve]], resolvingFunctions.[[Reject]] »).

So the behavior of `resolve` and `reject` comes from CreateResolvingFunctions. Following the link [there](https://tc39.es/ecma262/multipage/control-abstraction-objects.html#sec-createresolvingfunctions), we see that each of them has the following steps:
> 4. Let alreadyResolved be F.[[AlreadyResolved]].
> 5. If alreadyResolved.[[Value]] is true, return undefined.
> 6. Set alreadyResolved.[[Value]] to true.
> [… then the steps that actually reject or resolve the promise.]

So once a given promise has been settled one way or another, by either of those functions, subsequent calls to either one will just return void without doing anything.

[the remark this answers is not in the document]
Comment (chrisbobbe) [reply src/third/redux-persist/persistStore.js]: Cool, thanks for that explanation and fix! 🙂

**Comment by gnprice** [comment]
Thanks for the review! Merged, with that edit. (Actually I ended up using a `?:`, as the intended control flow is so simple.)

**Pull request merged**